feat(Tool Calling): Support `required` and specific function mode

CatherineSue opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

### Enforce tool calls with `tool_choice="required"` or specific function

Currently, **SGLang does not enforce tool calls** when:
- `tool_choice="required"` — where at least one tool function should be invoked
- `tool_choice={"type": "function", "function": {"name": "get_weather"}}` — where exactly one specific function must be used

Reference: https://platform.openai.com/docs/guides/function-calling?api-mode=chat#additional-configurations

In OpenAI-compatible behavior, the model is expected to **always call a function** in these cases. To support this, we must **constrain the model output**, which SGLang enables via the `--grammar-backend` flag (default: `xgrammar`).

Since `xgrammar` supports **EBNF-based constrained decoding**, we use this to strictly enforce tool call format and behavior.

> **Note**: The `outlines` backend does not support EBNF, but it provides strict constraints via JSON Schema. It can be added in the future, but is currently out of scope.

---

### Refactor `function_call_parser.py`

The existing `function_call_parser.py` is **over 1000 lines** and growing rapidly. As more model-specific formats are added, it becomes harder to maintain and test. This PR introduces modularity and simplifies its structure.

---

## Modifications

- 🔥 **Remove `MultiFormatParser`**
  - Supporting multiple tool call formats for a single model is unnecessary.
  - `FunctionCallParser.__init__` only accepts one `tool_call_parser` value — multiple detectors were dead code.
  - Simplifying to one active detector reduces parsing complexity and eliminates fragile merging logic.

- 🧱 **Introduce `EBNFComposer`**
  - Centralizes EBNF rule construction across detectors.
  - Enables consistent, strict grammar enforcement using XGrammar for `tool_choice="required"` and specific function calls.
  - Supports combining EBNFs from multiple formats when needed.

- 🛠️ **Fix `PythonicDetector.parse_streaming_increment`**
  - Return normal text when tool calls are not detected.
  - Handle partial tool calls across streamed chunks.
  - Properly manage buffer boundaries and normal text + tool call merging.
  - Fix edge cases in bracket-matching logic.

- 🧪 **Add unit tests**
  - New tests for tool call parsing using:
    - `tool_choice="required"`
    - specific function tool choices
  - Verify generated EBNF correctness via `test_function_call_parser.py`

- 🧹 **Testing cleanup**
  - Rename `test_function_calling.py` → `test_openai_function_calling.py`
  - Add `test_funcion_call_parser.py` for EBNF testing
<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

CatherineSue committed:
Refactor `function_call_parser.py`

- Remove `MultiFormatParser` as
  - It doesn't make sense for one model to have multiple types of tool call format.
  - Even it does happen, FunctionCallParser.__init__ only accept one tool_call_parser str, there is no way that there is gonna be multiple detectors.
  - Having multiple detectors significantly increase the logic of parsing results and setting up structured output constraints.

- Add `EBNFComposer`
  - Xgrammar backend can enforce model output without a trigger. With EBNF strings and utilizing Xgrammar, we can enforce model to use tools when tool_choice is ToolChoice specified or "required"

CatherineSue committed:
Move `function_call_parser.py` to function_call module

- `function_call_parser.py` has over 1000 lines. It will become more and more difficult to maintain

CatherineSue committed:
Fix PythonicDetector.parse_streaming_increment

- Add proper handling for text without tool calls by returning buffer as normal text
- Preserve and return normal text that appears before tool call markers
- Handle partial tool calls that span multiple streaming chunks
- Properly combine normal text with tool call results
- Fix edge cases in bracket detection and buffer management

CatherineSue committed:
Add missing imports

CatherineSue committed:
Format build_ebnf arguments in each Detector

CatherineSue committed:
Add Unittests

- Move `test_function_calling.py` to `test_openai_function_calling.py`
- Add test_funcion_call for sepecific tool_choice and required tool_choice
- Add `test_funcion_call_parser` to test EBNF strings

CatherineSue committed:
Simplify `build_ebnf` signature

CatherineSue committed:
Add doc: source for `json_grammar_ebnf_str`

zhyncs commented:
@bdubayah PTAL

CatherineSue committed:
Further constrain the type of arguments in EBNF str

CatherineSue committed:
Update UT

ispobock commented on python/sglang/srt/function_call/function_call_parser.py:
```diff
@@ -0,0 +1,169 @@
+        if tool_choice == "auto":
+            strict_tag = self.get_structure_tag()
+            return ("structural_tag", strict_tag)
+        else:
```
elif tool_choice == "required" or isinstance(tool_choice, ToolChoice)?

CatherineSue replied:
I updated the typing and the else here.
Technically, we filter none in adapter.py: https://github.com/sgl-project/sglang/blob/534d66f2f6021061494dc9d6162a58000600947f/python/sglang/srt/openai_api/adapter.py#L979-L992, but it is safe to filter here as well

ispobock commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -1156,20 +1158,24 @@ def v1_chat_generate_request(
-                sampling_params["structural_tag"] = convert_json_schema_to_str(
-                    strict_tag.model_dump(by_alias=True)
-                )
+                sampling_params[constraint_type] = constraint_value
```
Do we need to check the grammar backend here?

CatherineSue replied:
I think we don't need to check grammar backend here.
We have a pretty nice design for the constrained module. If a grammar backend doesn't support ebnf or any constraint, it will fall back to the BaseGrammarBackend and log a warning. See: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/constrained/base_grammar_backend.py#L130-L131

CatherineSue committed:
Update `get_structure_tag` to only use it when function.strict

CatherineSue committed:
Add more constraint to strict_tag

- We should not return strict_tag when there is not strict in the function. Otherwise it messes up with the normal model output.

CatherineSue committed:
Add more validation before get_structure_tag()

CatherineSue commented on python/sglang/srt/function_call/function_call_parser.py:
```diff
@@ -0,0 +1,169 @@
+            tool_structures.append(
+                StructuresResponseFormat(
+                    begin=info.begin,
+                    schema=schema,  # type: ignore
```
We should plan to add more UTs for the structure tag. There are probably more issues we haven't covered here, such as:
1. This schema won't work with PythonicDetector. Because it doesn't follow JSON schema.
cc @zhyncs

CatherineSue commented:
Placeholder for thorough tests
Let's **NOT** merge **until** this block is filled

1. tool_choice == "required" or ToolChoice for all model detectors
Llama4-Maverick (PythonicDetector)
Llama3.3-70B-Instruct (Llama32Detector)

3. tool_choice == "required" and parallel tool calls

CatherineSue committed:
Fix EBNF handling for enum

- enum should be wrapped differently based on type
- Add more docstring to EBNFComposer.build_ebnf

zhyncs committed:
Merge branch 'main' into chang/tool-call

CatherineSue committed:
Update function_calling notebook

CatherineSue committed:
Fix mistral_detector's missing eot_token

CatherineSue commented:
CI track:
[unittest-test-backend-8-gpu](https://github.com/sgl-project/sglang/actions/runs/15218602493/job/42810584447?pr=6550#logs) is failing due to eagle_worker.py
[unit-test-backend-1-gpu (1)](https://github.com/sgl-project/sglang/actions/runs/15218602493/job/42809692216?pr=6550#logs) is failing due to TestGemma3itServer
[performance-test-1-gpu-part-2](https://github.com/sgl-project/sglang/actions/runs/15218602493/job/42809692205?pr=6550#logs) is failing due to VLM online latency.

CatherineSue committed:
Fix test_function_call_parser

zhyncs committed:
Merge branch 'main' into chang/tool-call

CatherineSue committed:
Merge remote-tracking branch 'origin' into chang/tool-call

bdubayah commented on python/sglang/srt/function_call/pythonic_detector.py:
```diff
@@ -0,0 +1,141 @@
+
+        normal_text = self._buffer[:start] if start > 0 else ""
+
+        end = self._buffer.find("]", start)
```
This should be `rfind`, as you could have a list as a function argument

CatherineSue replied:
Good catch. I didn't consider it.

It seems using .rfind might cause issues when there are multiple complete tool calls in the buffer (an edge case maybe): e.g. `"[get_weather(location='NYC')] and [search(query='food')]"`

I introduced a bracket_count util method to count the index. 

Also added some uts.

bdubayah commented on python/sglang/srt/function_call/deepseekv3_detector.py:
```diff
@@ -0,0 +1,157 @@
+
+        if not hasattr(self, "_tool_indices"):
+            self._tool_indices = {
+                tool.function.name: i
```
Having one index for each function will cause issues if the model generates two tool calls for the same function

CatherineSue replied:
This block creates the tool_index from the original tools. So it should be fine.

bdubayah replied:
Suppose you have function calls produced like `[get_weather(city="New York"), get_weather(city="San Francisco")]`.  When you compute `tool_index=self._tool_indices.get(func_name, 0)` below, these will both have the same tool index, because the function name is the same. However, they are separate tool calls, so the index should be unique

CatherineSue replied:
Got it. I think this is an issue across the function call then. Let's create an issue to fix this later.

CatherineSue committed:
Enhance PythonicDetector to handle nested brackets in tool calls

- Add `_find_matching_bracket` to locate matching closing brackets, supporting nested structures.
- Add tests for handling partial nested bracket calls across multiple chunks.

CatherineSue commented:
![Screenshot 2025-05-23 at 5 38 39 PM](https://github.com/user-attachments/assets/1878e336-0e01-47b7-982a-a20aa927b1f5)

zhyncs committed:
Merge branch 'main' into chang/tool-call

Ubospica commented:
This PR looks great. Thanks @CatherineSue! 

The structural tag can be further improved to support the Pythonic tool calling. I think the current implementation is already very good, but in the future, a C++ version could be implemented on top of xgrammar for better performance.

Also, just to confirm, the current implementation of the required tool call doesn’t support mixed output of plain text and a function call, right? In other words, the output only includes a single tool call. I feel it's hard to use constrained decoding in implementing a combination of 1) required tool calling and 2) mixed output of text and tool call.

CatherineSue commented:
> This PR looks great. Thanks @CatherineSue!
> 
> The structural tag can be further improved to support the Pythonic tool calling. I think the current implementation is already very good, but in the future, a C++ version could be implemented on top of xgrammar for better performance.
> 
> Also, just to confirm, the current implementation of the required tool call doesn’t support mixed output of plain text and a function call, right? In other words, the output only includes a single tool call. I feel it's hard to use constrained decoding in implementing a combination of 1) required tool calling and 2) mixed output of text and tool call.

Thanks for all the advice @Ubospica !

Yes the current implementation doesn't support mixed output.
